### PR TITLE
Improve and document implementation of object_id

### PR DESCRIPTION
**Pull request by rberlind**
This fixes and documents the implementation of the new object_id attribute of the restapi resource. In particular, I replaced d.Id() with id on line 40 of common.go so that the actual value passed into object_id is set in the API object created by the call to NewAPIObject().

### CODE REVIEW

**Comment by rberlind** [comment]
I've added new id_header and id_header_is_url keys on the provider that support retrieving the object ID from a specified header returned by the POST method used to create the object. If id_header_is_url is set to true, then the ID will be set to the last segment of the URL using "/" as a separator and removing any final "/".  Otherwise, if id_header_is_url is false, then the ID will be set to the entire header.  These additions are needed by APIs like the vRO REST API that return a URL in the Location header that would look like "https://vroserver.xyz.com:8281/vco/api/workflows/06db9b35-bc27-4895-a878-942144d62790/executions/8ab79b96634d0d210163cfe7dd684401/". I added code to support these in the obvious places.

In order to support this functionality, I've modified send_request() in api_client.go to return both the response headers and body.  I currently am returning the headers as an http.Header type rather than a generic map because this was easier.  But it could be construed as an odd choice.  Eventually, this could be modified. I also updated all calls into send_request() to reflect the change. Additionally, in the functions that call it, I changed the res_str string to res_body to better differentiate it from the new res_headers object. I also added debug prints for the headers, not so much because I wanted them but because `go build` complained when I did not use the res_headers object that send_request() was returning.

The processing of the headers to find one matching id_header is done in create_object() of api_object.go.

I tested that I could read a header using the Vault API in a strange sort of way, but have not yet tested with the vRO API.  I hope the customer I'm working with will do that today.  I am confident that the processing of non-URL headers works.

I did not run any of the tests, but I did modify them to use updated signatures of any functions for which I modified parameters.  So, my changes should not have broken the tests.  Ideally, they should be modified to actually test the new keys.

I also fixed a problem with the processing of non-existent *_path keys in api_object.go, replacing the checks that tested for `nil` to instead test for `""`.  This is needed since the ResourceData.Get() function only returns nil if the key is not in the schema; otherwise, it returns the default value of the type, so "" for strings.  See https://godoc.org/github.com/hashicorp/terraform/helper/schema#ResourceData.Get.

I modified the test for copy_keys in update_state() of api_object.go to use `if len(obj.api_client.copy_keys) > 0  && obj.api_data != nil` since it seems to me that we should be checking both that there are keys that the user wants to copy and that the obj.api_data object is not empty. Beyond that check, I felt it is up to the user to know if their api_data will contain the keys they want to copy.

I hope the changes I have made are ok.  Since I'm not really a developer and had to figure out how to do various things in Go, my code choices are probably not the best.  I apologize for that.

File: restapi/common.go [hunk 1]
```diff
@@ -37,7 +37,7 @@ func make_api_object(d *schema.ResourceData, m interface{}) (*api_object, error)
     post_path,
     put_path,
     delete_path,
-    d.Id(),
+    id,
```
Comment (DRuggeri) [inline restapi/common.go]: Argh - thanks for the fix. I have merged this commit into master and will produce a release with the fix!

**Comment by rberlind** [comment]
Before you create a new release, I might have another commit.
My customer did testing against vRO this morning with some success.  We were able to extract the header which was a URL and parse out the final segment of the URL as the ID.  However, I just determined that my current handling of the id_header is only done if create_returns_object is true.  That had actually been intentional as I was writing this yesterday, but I now realize that the vRO API is only returning headers and nothing in the body of the response.  So, in this case, I actually DO want the provider to do the GET.  So, I will change to do my processing even if create_returns_object is false.

**Comment by DRuggeri** [comment]
Let's take a pause here. Thinking more about HTTP semantics, maybe it just makes more sense to follow the Location header as a redirect, rather than trying to parse it as data. I don't know much about this API, but it seems like it is returning a reference to the data that was gathered as a redirect - I would expect a subsequent GET to the URL in `Location` would give the data. Is that the case? What do you think about doing that instead?

**Comment by DRuggeri** [comment]
Also, I'll go ahead and roll a release with the fix for `object_id` while we figure out the best way to handle this particular case.

**Comment by rberlind** [comment]
Thanks.  I just pushed another commit to my fork that now allows the id_header attribute to be used with create_returns_object = false.  I've tested that against the Vault API and now see the read being done after the create as desired.  I will build a Linux binary for my customer and have them test tomorrow morning against vRO.  (They're in Europe.) This will help validate in a more real-world setting than my tests that the object ID can be read from a URL in a header returned by the POST and that the read can then be done against the correct path using that ID.

**Comment by rberlind** [comment]
Also updated the README.md to reflect the changes.

**Comment by DRuggeri** [comment]
OK - but did you see my question about how this API behaves? To be clear, I think parsing headers like this is very API-specific. If this API uses Location with "/" delimeters, another may choose to use RFC 5988 Link headers which don't conform with that at all.

What I *suspect* is probably going on with this particular API is that, in response to the POST of the data, it is returning a 301 or a 302 to where you can go to GET the information. I had actually started writing code to handle the case of redirects in `api_client.go`, but never quite completed it. I would much rather implement some form of handling of normal HTTP behaviors rather than API-specific things.

**Comment by rberlind** [comment]
Apologies for not responding to your question, Daniel. I ran out of time to respond before the weekend.  Here is the link to the vRO API doc that mentions that the URL with the execution ID will be in the Location header: https://docs.vmware.com/en/vRealize-Orchestrator/7.2/com.vmware.vrealize.orchestrator-develop-web-services.doc/GUID-5312E7E5-90B7-4D14-9BDA-43BF22E0DDF4.html.  Specifically, it says "If the POST request is successful, you receive the status code 202 with an empty response body and a link to the newly created execution object in the Location header."

Note that the Location header will be set to something like "https://itsusabsdvro01d.jnj.com:8281/vco/api/workflows/06db9b35-bc27-4895-a878-942144d62790/executions/8ab79b96634d0d210163cfe7dd684401/".  For purposes of tracking the created execution against the workflow, the last part of this URL is the ID that should be associated with the object created by the restapi resource.  So, I don't think re-directing will help.

But if you prefer to not include my modified code for the id_header, that is ok.  I have been able to provide my customer with binaries that have my modified code.

**Pull request closed without merging**